### PR TITLE
Initialze Daru::Index from range

### PR DESCRIPTION
**Pull request by lokeshh**

### CODE REVIEW

File: lib/daru/index.rb [hunk 1]
```diff
@@ -24,7 +24,7 @@ def self.new *args, &block
       source = args[0]
 
       idx =
-      if source and source[0].is_a?(Array)
+      if source.is_a?(Array) and source[0].is_a?(Array)
```
Comment (v0dro) [inline lib/daru/index.rb]: why introduce this conditional?

File: spec/index_spec.rb [hunk 2]
```diff
@@ -37,6 +37,12 @@
       expect(idx.to_a).to eq(['speaker', 'mic', 'guitar', 'amp'])
     end
 
+    it "creates an Index from Range" do
+      idx = Daru::Index.new 1..5
+
+      expect(idx).to eq(Daru::Index.new [1, 2, 3, 4, 5])
+    end
```
Comment (v0dro) [inline spec/index_spec.rb]: everything else is fine.

File: lib/daru/index.rb [hunk 3]
```diff
@@ -51,8 +51,8 @@ def map(&block)
 
     def initialize index
       index = 0                         if index.nil?
-      index = Array.new(index) { |i| i} if index.is_a? Integer
-      index = index.to_a                if index.is_a? Daru::Index
+      index = index.times.to_a          if index.is_a? Integer
+      index = index.to_a                if index.is_a? Enumerable
```
Comment (zverok) [inline lib/daru/index.rb]: Maybe `case` will be cleaner here?

File: lib/daru/index.rb [hunk 4]
```diff
@@ -51,8 +51,8 @@ def map(&block)
...
 
       @relation_hash = {}
       index.each_with_index do |n, idx|
```
Comment (zverok) [inline lib/daru/index.rb]: BTW, as far as I can see, the code of this each_with_index could be replaces with just `@relations_hash = index.each_with_index.to_h`

File: lib/daru/index.rb (hunk 1 above)
Comment (lokeshh) [inline lib/daru/index.rb]: Range is not an array, so source[0] was producing an error.

Comment (zverok) [inline lib/daru/index.rb]: `source.respond_to?(:first) && source.first.is_a?(Array)` maybe?..

File: lib/daru/index.rb [hunk 5]
```diff
@@ -50,15 +50,17 @@ def map(&block)
...
-      index.each_with_index do |n, idx|
-        @relation_hash[n] = idx 
+      index = case
+        when index.nil?
+          []
+        when index.is_a?(Integer)
+          index.times.to_a
+        when index.is_a?(Enumerable)
+          index.to_a
       end
```
Comment (zverok) [inline lib/daru/index.rb]:
```ruby
index = case index
  when nil
  when Integer
  when Enumerable
  else
  end
```

File: lib/daru/index.rb (hunk 1 above)
Comment (lokeshh) [inline lib/daru/index.rb]: Sure. It shows the intent better. Thanks

File: lib/daru/index.rb (hunk 5 above)
Comment (lokeshh) [inline lib/daru/index.rb]: Beautiful!

File: lib/daru/index.rb [hunk 6]
```diff
@@ -50,15 +50,17 @@ def map(&block)
...
-      @relation_hash = {}
-      index.each_with_index do |n, idx|
-        @relation_hash[n] = idx 
+      index = case index
+        when nil
+          []
+        when Integer
+          index.times.to_a
+        when Enumerable
+          index.to_a
```
Comment (zverok) [inline lib/daru/index.rb]: Sorry for being pedantic, but I think we still need `else` here, or we have pretty cryptic error:
```ruby
Daru::Index.new('foo')
# => NoMethodError: undefined method `each_with_index' for nil:NilClass
```

Comment (zverok) [inline lib/daru/index.rb]: (And spec for this case would be great also)

File: lib/daru/index.rb [hunk 7]
```diff
@@ -50,15 +50,17 @@ def map(&block)
...
       end
 
+      @relation_hash = index.each_with_index.to_h
+
```
Comment (zverok) [inline lib/daru/index.rb]: Now we can add `.freeze` (from line 64) to the end of this statement, I think.

File: lib/daru/index.rb (hunk 6 above)
Comment (lokeshh) [inline lib/daru/index.rb]: No, it's good. Keep telling me more places where it can be improved. Thanks!

File: lib/daru/index.rb [hunk 8]
```diff
@@ -24,7 +24,7 @@ def self.new *args, &block
       source = args[0]
 
       idx =
-      if source and source[0].is_a?(Array)
+      if source.respond_to?(:first) && source.first.is_a?(Array)
```
Comment (v0dro) [inline lib/daru/index.rb]: I still don't understand the reason behind replacing the previous line with this one. Also, don't squash commits until the PR is ready. Once the commit is gone, the comments made on it cannot be seen.

**Comment by lokeshh** [comment]
@v0dro This is ready to be merged.

**Comment by lokeshh** [comment]
@v0dro The `to_h` works to seem fine in Daru even when using Ruby 2.0. I am currently looking at what the problem is.

**Comment by lokeshh** [comment]
This is weird:

```ruby
[4] pry(main)> RUBY_VERSION
=> "2.0.0"
[5] pry(main)> a = [1, 2, 3].each_with_index
=> #<Enumerator: ...>
[6] pry(main)> a.method(:to_h).source_location
=> ["/home/ubuntu/.rvm/gems/ruby-2.0.0-p643/gems/backports-3.6.8/lib/backports/2.1.0/enumerable/to_h.rb", 4]
```

**Comment by lokeshh** [comment]
Is there any way to check Daru version inside pry?

**Comment by lokeshh** [comment]
Finally, here the fix:

Adding `  spec.add_development_dependency 'nmatrix', '~> 0.2.1'` to specfile of statsample-glm solves the issue.

**Comment by v0dro** [comment]
Why was this happening?

**Comment by lokeshh** [comment]
I am investigating.

**Comment by lokeshh** [comment]
Here's a permanent fix and it explains the issue: https://github.com/v0dro/daru/pull/126

**Pull request merged**